### DISCUSSION TITLE
Multiple Rust modules with "cargo test" (minimal example included)

### DISCUSSION THREAD

**Discussion by SanderVocke** [KDAB/cxx-qt#1014] [Q&A] [answered]
Hi there,

I am starting to get the hang of using `cxx-qt` and would love to use it to slowly migrate a QML project's custom QML types from PySide to cxx-qt.

In my approach I compile the cxx-qt QObject types into a static library, then link that into a dynamic library with a registration function that registers them through `qmlRegisterType`.

It seems to work nicely.

However, I also would like to make use of `cargo test` to unit-test the objects on the Rust side individually. For my first module, this worked fine. But when creating more than one Rust module in the same crate, I run into linker errors.

To illustrate I made this minimal example: https://github.com/SanderVocke/cxx-qt-multi-module-problem

It is quite straightforward - basically it is the minimal-qml example, plus:
- a single C++ header for creating unique pointers, to use in unit-testing;
- an extra module added;
- each module exposes test functions with Cargo's test macros.

`cargo build` works fine, but `cargo test` runs into linker issues. The README gives more detail. This is on 0.6.1.

Would you have any idea how to modify or restructure this? Ideally I would still like it to be a single crate, because I expect to make many QObject derivatives and a crate per object would be overkill.

Thanks in any case!

P.S.: if with a small modification this structure can be made to work, I would suggest adding something like this into the `examples` folder.

**Comment by ahayzen-kdab** [comment, marked as answer]
It could be that both modules are outputting to the same header file?

Could you try using `cxx_file_stem` to specify different files in the bridge macro?

See the book for more info https://kdab.github.io/cxx-qt/book/bridge/index.html#cxx_file_stem

**Reply by LeonMatthesKDAB** [reply]
Just tested it locally: This solves the issue for me!
We probably need a better error message for this!

**Reply by LeonMatthesKDAB** [reply]
At least with the newest version of CXX-Qt (I bumped it to the version in our git repo)

**Reply by ahayzen-kdab** [reply]
Yeah, maybe the new build system can detect that, but it's figuring out if the existing file is from a previous build or if it's from a another module.

Or might just be that CxxQtBuilder needs to keep track of the output files and detect duplicates etc.

**Reply by SanderVocke** [reply]
Ah, that's probably it. I didn't realize the module name influenced this. Will try this, thanks!

**Reply by SanderVocke** [reply]
That's it, thanks! Each bridge module name should be different.
